# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [1]:
print('Escola de Dados da Alura!')

Escola de Dados da Alura!


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [2]:
print('Nome: Débora\nSobrenome: Carvalho')

Nome: Débora
Sobrenome: Carvalho


### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


In [3]:
print('D\nE\nB\nO\nR\nA')

D
E
B
O
R
A


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [7]:
print(13, 'maio', 2003)

13 maio 2003


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [ ]:
print('Ano atual: 2024')

## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [11]:
nome = str(input('Digite o seu nome: '))
print(f'Olá, {nome}!')

Olá, Debora!


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [13]:
nome = str(input('Digite o seu nome: '))
idade = int(input('Digite a sua idade: '))
print(f'Olá, {nome}, você tem {idade} anos.')

Olá, Debora, você tem 12 anos.


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [14]:
nome = str(input('Digite o seu nome: '))
idade = int(input('Digite a sua idade: '))
altura = float(input('Digite sua altura (em metros)'))
print(f'Olá, {nome}, você tem {idade} anos e mede {altura} metros!')

Olá, Debora, você tem 21 anos e mede 1.62 metros!


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [15]:
n1 = float(input('Digite o primeiro valor: '))
n2 = float(input('Digite o segundo valor: '))
print(n1+n2)

8.6


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [16]:
n1 = float(input('Digite o primeiro valor: '))
n2 = float(input('Digite o segundo valor: '))
n3 = float(input('Digite o terceiro valor: '))
print(n1+n2+n3)

59.5


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [18]:
n1 = float(input('Digite o primeiro valor: '))
n2 = float(input('Digite o segundo valor: '))
print(n1-n2)

8.2


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [19]:
n1 = float(input('Digite o primeiro valor: '))
n2 = float(input('Digite o segundo valor: '))
print(n1*n2)

18.0


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [ ]:
numerador = float(input('Digite o numerador: '))
denominador = float(input('Digite o denominador (deve ser diferente de 0): '))

if denominador == 0:
    print('Erro: O denominador não pode ser 0.')
else:
    print(numerador / denominador)

1.2


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [28]:
base = float(input('Digite a base: '))
expoente = float(input('Digite o expoente: '))
print(base ** expoente)

36.0


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [31]:
numerador = int(input('Digite o numerador: '))
denominador = int(input('Digite o denominador (deve ser diferente de 0): '))

if denominador == 0:
    print('Erro: O denominador não pode ser 0.')
else:
    print(int(numerador / denominador))

4


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [34]:
numerador = int(input('Digite o numerador: '))
denominador = int(input('Digite o denominador (deve ser diferente de 0): '))

if denominador == 0:
    print('Erro: O denominador não pode ser 0.')
else:
    print(numerador % denominador)

1


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [35]:
n1 = float(input('Digite a nota 1: '))
n2 = float(input('Digite a nota 2: '))
n3 = float(input('Digite a nota 3: '))
media = (n1+n2+n3)/3
print(media)

7.866666666666667


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [38]:
n1 = float(input('Digite o primeiro número: '))
n2 = float(input('Digite o segundo número: '))
n3 = float(input('Digite o terceiro número: '))
n4 = float(input('Digite o quarto número: '))
media = (n1*1 + n2*2 + n3*3 + n4*4) / (1 + 2 + 3 + 4)
print(media)

3.0


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [39]:
frase = str('Olá amigos!')
print(frase)

Olá amigos!


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [40]:
frase = str(input('Digite uma frase: '))
print(frase)

Boas festas!


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [41]:
frase = str(input('Digite uma frase: '))
print(frase.upper())

OLA


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [42]:
frase = str(input('Digite uma frase: '))
print(frase.lower())

ola amigos!


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [43]:
frase = '  Seja bem-vindo ao nosso bairro! '
print(frase.strip())

Seja bem-vindo ao nosso bairro!


#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [44]:
frase = str(input('Digite uma frase: '))
print(frase.strip())

Desculpe-me, estou ocupada agora.


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [46]:
frase = str(input('Digite uma frase: '))
print(frase.strip().upper())

ESTOU ESCREVENDO ESSA FRASE EM LETRAS MINÚSCULAS.


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [48]:
frase = str(input('Digite uma frase: '))
print(frase.replace('e', 'f'))

Estou fspfrando flf ainda hojf.


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [49]:
frase = str(input('Digite uma frase: '))
print(frase.replace('a', '@'))

Doces @mor@s for@m colhid@s próxim@s @ ess@s árvores.


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [50]:
frase = str(input('Digite uma frase: '))
print(frase.replace('s', '$'))

Sua$ palavra$ $ão muito $ábia$.


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [54]:
n1 = float(input('Digite o primeiro número: '))
n2 = float(input('Digite o segundo número: '))

if n1>n2:
    print(f'O número maior é {n1}.')
elif n1==n2:
    print(f'Os números são iguais.')
else:
    print(f'O número maior é {n2}.')

O número maior é 9.0.


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [57]:
pCrescimento = float(input('Digite o percentual de crescimento da produção: '))

if pCrescimento > 0:
    print('Houve crescimento na produção da empresa.')
elif pCrescimento == 0:
    print('Percentual de crescimento nulo.')
else:
    print('Houve decrescimento na produção da empresa.')

Houve decrescimento na produção da empresa.


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [72]:
letra = str(input('Digite uma letra: '))

if letra == 'a' or letra == 'e' or letra == 'i' or letra == 'o' or letra == 'u':
    print('É uma vogal')
else:
    print('É uma consoante.')

É uma consoante.


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [77]:
v1 = float(input('Digite o valor médio do ano 1: '))
v2 = float(input('Digite o valor médio do ano 2: '))
v3 = float(input('Digite o valor médio do ano 3: '))

if v1 > v2 and v1 > v3:
    valorAlto = v1
elif v2 > v1 and v2 > v3:
    valorAlto = v2
else:
    valorAlto = v3

if v1 < v2 and v1 < v3:
    valorBaixo = v1
elif v2 < v1 and v2 < v3:
    valorBaixo = v2
else:
    valorBaixo = v3    

print(f'O valor mais alto entre esses três anos para o modelo do carro é {valorAlto}.\nJá o valor mais baixo é {valorBaixo}.')

O valor mais alto entre esses três anos para o modelo do carro é 158.0.
Já o valor mais baixo é 12.0.


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [80]:
p1 = float(input('Digite o preço do produto 1: '))
p2 = float(input('Digite o preço do produto 2: '))
p3 = float(input('Digite o preço do produto 3: '))

if p1 < p2 and p1 < p3:
    menorPreco = p1 
    produto = 1
elif p2 < p1 and p2 < p3:
    menorPreco = p2
    produto = 2
else:
    menorPreco = p3
    produto = 3

print(f'O produto {produto} é o mais barato com preço de {menorPreco}.')

O produto 3 é o mais barato com preço de 18.0.


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [81]:
num1 = float(input("Digite o primeiro número: "))
num2 = float(input("Digite o segundo número: "))
num3 = float(input("Digite o terceiro número: "))

if num1 >= num2 and num1 >= num3:
    primeiro = num1
    if num2 >= num3:
        segundo = num2
        terceiro = num3
    else:
        segundo = num3
        terceiro = num2
elif num2 >= num1 and num2 >= num3:
    primeiro = num2
    if num1 >= num3:
        segundo = num1
        terceiro = num3
    else:
        segundo = num3
        terceiro = num1
else:
    primeiro = num3
    if num1 >= num2:
        segundo = num1
        terceiro = num2
    else:
        segundo = num2
        terceiro = num1

print(f"Os números em ordem decrescente são: {primeiro}, {segundo}, {terceiro}")


Os números em ordem decrescente são: 5.0, 2.0, 1.0


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [89]:
turno = str(input('Em que turno você estuda? Responda com manhã, tarde ou noite: '))

if turno == 'manhã' or turno == 'Manhã':
    print('Bom dia!')
elif turno == 'tarde' or turno == 'Tarde':
    print('Boa tarde!')
elif turno == 'noite' or turno == 'Noite':
    print('Boa noite!')
else:
    print('Valor Inválido!')

Bom dia!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [83]:
num = int(input("Digite um número: "))
resto = num % 2

if resto == 0:
    print(f'O número {num} é par.')
else:
    print(f'O número {num} é impar.')

O número 3 é impar.


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [93]:
num = float(input('Digite um número: '))

if num.is_integer():
    print('Este número é inteiro.')
else:
    print('Este número é decimal.')

Este número é inteiro.


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [3]:
n1 = float(input('Digite o primeiro número: '))
n2 = float(input('Digite o segundo número: '))
op = int(input('Digite o número correspondente à operação desejada:\n1-Adição\n2-Subtração\n3-Multiplicação\n4-Divisão\n5-Exponenciação\n6-Média\n0-Sair'))

# operações
if op == 1:
    resultado = n1 + n2
    operacao = 'soma'
elif op == 2:
    resultado = n1 - n2
    operacao = 'subtração'
elif op == 3:
    resultado = n1 * n2
    operacao = 'multiplicação'
elif op == 4:
    if n2 != 0:
        resultado = n1 / n2
        operacao = 'divisão'
    else:
        print("Erro: Divisão por zero não é permitida.")
elif op == 5:
    resultado = n1 ** n2
    operacao = 'exponenciação'
elif op == 6:
    resultado = (n1 + n2)/2
    operacao = 'média'

# verificar se o número é par ou ímpar 
if (resultado % 2) == 0:
    numParImpar = 'par' 
elif (resultado % 2) != 0:
    numParImpar = 'ímpar' 

# verificar se o número é positivo ou negativo
if resultado < 0:
    numPosNeg = 'negativo'
elif resultado == 0:
    numPosNeg = 'nulo'
else:
    numPosNeg = 'positivo'

# verificar se o número é inteiro ou decimal
if resultado.is_integer():
    numIntDec = 'inteiro'
else:
    numIntDec = 'decimal'

# exibir o resultado
print(f'O resultado da {operacao} entre {n1} e {n2} é {resultado}. Este número é {numParImpar}, {numPosNeg} e {numIntDec}.')

Erro: Divisão por zero não é permitida.
O resultado da soma entre 6.0 e 0.0 é 8.7. Este número é ímpar, positivo e decimal.


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dos lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dos lados iguais;
- Triângulo Escaleno: três lados diferentes;

In [ ]:
lado1 = float(input('Digite a medida do primeiro lado para o triângulo: '))
lado2 = float(input('Digite a medida do segundo lado para o triângulo: '))
lado3 = float(input('Digite a medida do terceiro lado para o triângulo: '))

# verificar se pode formar um triângulo
if (lado1 + lado2) > lado3:
    triangulo = True
elif (lado2 + lado3) > lado1:
    triangulo = True
elif (lado1 + lado3) > lado2:
    triangulo = True 
elif lado1==0 or lado2==0 or lado3==0:
    triangulo = False
else:
    triangulo = False

# verificar o tipo de triângulo
if triangulo == True:
    if lado1 == lado2 == lado3:
        tipoT = 'equilátero'
    elif lado1 == lado2 or lado2 == lado3 or lado1 == lado3:
        tipoT = 'isósceles'
    elif lado1 != lado2 != lado3:
        tipoT = 'escaleno'
    print(f'Com as medidas {lado1}, {lado2} e {lado3} para os lados, é possível formar um triângulo do tipo {tipoT}.')
elif triangulo == False:
    print(f'Com as medidas {lado1}, {lado2} e {lado3} para os lados, não é possível formar um triângulo.')

Com as medidas 0.0, 0.0 e 6.0 para os lados, é possível formar um triângulo do tipo isósceles.


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [17]:
quantL = float(input('Informe a quantidade de litros vendidos: '))
tipo = str(input('Informe o tipo de combustível (digite E para etanol e D para diesel): '))

# verificar o preço do tipo de combustível
if tipo == 'E':
    precoL = 1.70
elif tipo == 'D':
    precoL = 2.00

# verificar o percentual de desconto do tipo de combustível
if tipo == 'E':
    if quantL <= 15.0:
        pDesconto = 0.02
    else:
        pDesconto = 0.04
elif tipo == 'D':
    if quantL <= 15.0:
        pDesconto = 0.03
    else:
        pDesconto = 0.05

valorDesconto = precoL * quantL * pDesconto
valorTotal = (precoL * quantL) - valorDesconto

print(f'O total a ser pago é R$ {valorTotal:.2f}.')

O total a ser pago é R$ 26.11.


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [21]:
vendasAno1 = int(input('Informe a quantidade de vendas do ano de 2022: '))
vendasAno2 = int(input('Informe a quantidade de vendas do ano de 2023: '))

# cálculo de variação percentual 
variacao_percentual = ((vendasAno2 - vendasAno1) / vendasAno1)  * 100  

# exibir sugestões para tomada de decisão de acordo com a variação percentual
if variacao_percentual > 20:
    print(f'Ocorreu uma variação de {variacao_percentual:.2f}% entre as vendas de 2022 e 2023. Sugere-se efetuar uma bonificação para o time de vendas.')
elif variacao_percentual <= 20 and variacao_percentual > 2:
    print(f'Ocorreu uma variação de {variacao_percentual:.2f}% entre as vendas de 2022 e 2023. Sugere-se efetuar uma pequena bonificação para o time de vendas.')
elif variacao_percentual <= 2 and variacao_percentual >= (-10):
    print(f'Ocorreu uma variação de {variacao_percentual:.2f}% entre as vendas de 2022 e 2023. Sugere-se efetuar um planejamento de políticas de incentivo às vendas.')
elif variacao_percentual < (-10):
    print(f'Ocorreu uma variação de {variacao_percentual:.2f}% entre as vendas de 2022 e 2023. Sugere-se efetuar um corte de gastos.')

Ocorreu uma variação de 0.50% entre as vendas de 2022 e 2023. Sugere-se efetuar um planejamento de políticas de incentivo às vendas.


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

In [27]:
n1 = int(input('Digite o 1° número inteiro: '))
n2 = int(input('Digite o 2° número inteiro: '))

if n1 < n2:
    i = n1
    while i <= n2:
        print(i)
        i += 1
elif n1 > n2:
    i = n1
    while i >= n2:
        print(i)
        i -= 1
else:
    print(f"Os números são iguais: {n1}")


Os números são iguais: 6


#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

In [29]:
elementosA = 4
elementosB = 10
taxaA = 0.03
taxaB = 0.015
dias = 0

while elementosA < elementosB:
    elementosA += elementosA * taxaA
    elementosB += elementosB * taxaB
    dias += 1

print(f"A colônia A ultrapassará ou igualará a colônia B em {dias} dias.")
print(f"População final da colônia A: {elementosA:.2f}")
print(f"População final da colônia B: {elementosB:.2f}")

A colônia A ultrapassará ou igualará a colônia B em 63 dias.
População final da colônia A: 25.75
População final da colônia B: 25.55


#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

In [ ]:
# inicializando a variável de contagem
avaliacao = 1
# a variável abaixo irá guardar a quantidade de usuários que responderão à pesquisa. No caso deste exercício, seria 15
totalAvaliacao = int(input('Insira a quantidade de usuários que responderão à pesquisa: '))

while avaliacao <= totalAvaliacao:
    nota = int(input('Avalie o serviço com um número de 0 a 5: '))

    if nota < 0 or nota > 5:
        print(f'Nota inválida (você inseriu {nota}). Para avaliar, insira um número entre 0 e 5.')
    else:
        print(f'Avaliação n°{avaliacao} enviada com sucesso. Nota: {nota}')
        avaliacao += 1

Avaliação n°1 enviada com sucesso. Nota: 0
Avaliação n°2 enviada com sucesso. Nota: 2
Nota inválida (você inseriu 9). Para avaliar, insira um número entre 0 e 5.
Avaliação n°3 enviada com sucesso. Nota: 2


#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

In [7]:
# inicializando as variáveis
soma_temperaturas = 0
contador = 0

print("Digite as temperaturas em Celsius (digite -273 para encerrar):")

# laço para leitura das temperaturas
while True:
    temperatura = float(input(f"Temperatura {contador + 1}: "))
    
    # verifica se é o valor para encerrar
    if temperatura == -273:
        break
    
    # acumula as temperaturas e incrementa o contador
    soma_temperaturas += temperatura
    contador += 1

# calcula e exibe a média, se houverem temperaturas válidas
if contador > 0:
    media = soma_temperaturas / contador
    print(f"A média das temperaturas informadas é {media:.2f}°C.")
else:
    print("Nenhuma temperatura válida foi informada.")

Digite as temperaturas em Celsius (digite -273 para encerrar):
A média das temperaturas informadas é 72.25°C.


#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

In [ ]:
num = int(input('Digite um número inteiro para calcular seu fatorial: '))
fatorial = 1

for i in range(1,(num+1),1):
    fatorial = fatorial * i   # multiplica o valor atual de fatorial pelo índice

print(f'O fatorial de {num} é {fatorial}.')

O fatorial de 5 é 120.


### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

In [16]:
num = int(input('Digite um número inteiro de 1 a 10 para calcular sua tabuada: '))

print(f'Tabuada do {num}:')

for i in range(1,11,1):
    print(f'{num} x {i} = {num*i}')

Tabuada do 3:
3 x 1 = 3
3 x 2 = 6
3 x 3 = 9
3 x 4 = 12
3 x 5 = 15
3 x 6 = 18
3 x 7 = 21
3 x 8 = 24
3 x 9 = 27
3 x 10 = 30


#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
``` 
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
